legate/core/machine.py

```diff
@@ -0,0 +1,262 @@
+
+
+# Make this consistent with TaskTarget in mapping.h
+@unique
+class ProcessorKind(IntEnum):
+    GPU = 1
+    OMP = 2
+    CPU = 3
+
+
+ProcKindLike = Union[str, ProcessorKind]
+
+
+def sanitize_kind(kind: ProcKindLike) -> ProcessorKind:
+    try:
+        return ProcessorKind[kind.upper()] if isinstance(kind, str) else kind
+    except KeyError:
+        raise ValueError(f"Invalid processor kind: {kind}")
+
+
+# Inclusive range of processor ids
+@dataclass(frozen=True)
+class ProcessorRange:
+    # the kind is being used just in compatibility checks
+    kind: ProcessorKind
+    per_node_count: int
+    lo: int
+    hi: int
+
+    @staticmethod
+    def create(
+        kind: ProcessorKind, per_node_count: int, lo: int, hi: int
+    ) -> ProcessorRange:
+        if hi < lo:
+            lo = 1
+            hi = 0
+        return ProcessorRange(kind, per_node_count, lo, hi)
+
+    @property
+    def empty(self) -> bool:
+        return self.hi < self.lo
+
+    def __len__(self) -> int:
+        return self.hi - self.lo + 1
+
+    def __and__(self, other: ProcessorRange) -> ProcessorRange:
+        if self.kind != other.kind:
+            raise ValueError(
+                "Intersection between different processor kinds: "
+                f"{self.kind.name} and {other.kind.name}"
+            )
+        assert self.per_node_count == other.per_node_count
+        return ProcessorRange.create(
+            self.kind,
+            self.per_node_count,
+            max(self.lo, other.lo),
+            min(self.hi, other.hi),
+        )
+
+    def slice(self, sl: slice) -> ProcessorRange:
+        if sl.step is not None and sl.step != 1:
+            raise ValueError("The slicing step must be 1")
+        sz = len(self)
+        new_lo = self.lo
+        new_hi = self.hi
+        if sl.start is not None:
+            if sl.start >= 0:
+                new_lo += sl.start
+            else:
+                new_lo += max(0, sl.start + sz)
+        if sl.stop is not None:
+            if sl.stop >= 0:
+                new_hi = self.lo + sl.stop - 1
+            else:
+                new_hi = self.lo + max(0, sl.stop + sz)
+
+        return ProcessorRange.create(
+            self.kind, self.per_node_count, new_lo, new_hi
+        )
+
+    def get_node_range(self) -> tuple[int, int]:
+        if self.empty:
+            raise ValueError(
+                "Illegal to get a node range of an empty processor range"
+            )
+        return (self.lo // self.per_node_count, self.hi // self.per_node_count)
+
+    def __repr__(self) -> str:
+        if self.hi < self.lo:
+            return "<empty>"
+        else:
+            return f"[{self.lo}, {self.hi}] ({self.per_node_count} per node)"
+
+    def pack(self, buf: BufferBuilder) -> None:
+        buf.pack_32bit_uint(self.per_node_count)
+        buf.pack_32bit_uint(self.lo)
+        buf.pack_32bit_uint(self.hi)
+
+
+ProcSlice = Tuple[ProcKindLike, slice]
+
+
+class Machine:
```

Comment:
I agree that a proper documentation on the machine and scoping would be useful.

> Is Machine the right name? I wonder if something like ProcessorGroup is more accurate.  Machine carries a connotation of being the whole machine or a single node.

Though we only have processors in `Machine` today,  we will later want to use it to keep track of other types of resources. One eminent addition to `Machine` would be memories; for example, you could potentially want to control the amount of memory to assign to each section of the code. So, I wouldn't name it specifically for processors purely because it only has processors today.